pkg/controller/deployment/deployment_controller.go

```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	}
 	// Check if we can scale down.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	maxScaledDown := allPodsCount - minAvailable
+
+	// Clean up  unhealthy replicas first, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment, maxScaledDown)
+	if err != nil {
+		return false, nil
+	}
+
+	totalScaledDown := cleanupCount
+	if cleanupCount < maxScaledDown {
+		// Scale down old rcs, need check maxUnavailable to ensure we can scale down
+		scaledDownCount, err := dc.scaleDownOldRCsForRollingUpdate(allRCs, oldRCs, deployment, maxScaledDown-cleanupCount)
+		if err != nil {
+			return false, nil
+		}
+		totalScaledDown = cleanupCount + scaledDownCount
+	}
+
+	if expectationsCheck {
+		dc.podExpectations.ExpectDeletions(dKey, totalScaledDown)
+	}
+
+	return totalScaledDown > 0, nil
+}
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxCleanupCount int) (int, error) {
 	minReadySeconds := deployment.Spec.Strategy.RollingUpdate.MinReadySeconds
-	// Check the expectations of deployment before counting available pods
-	dKey, err := controller.KeyFunc(&deployment)
+	// Safely scale down all old rcs with unhealthy replicas. ReplicationController/ReplicaSet will sort the pods in the order
+	// such that not-ready < ready, unscheduled < scheduled, and pending < running. This ensures that unhealthy replicas will
+	// been deleted first and won't increase unavailability.
+	totalScaledDown := 0
+	for _, targetRC := range oldRCs {
+		if totalScaledDown >= maxCleanupCount {
+			break
+		}
+
+		if targetRC.Spec.Replicas == 0 {
+			// cannot scale down this RC.
+			continue
+		}
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, minReadySeconds)
+		if err != nil {
+			return totalScaledDown, fmt.Errorf("could not find available pods: %v", err)
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		newReplicasCount := readyPodCount
+		if totalScaledDown+(targetRC.Spec.Replicas-newReplicasCount) > maxCleanupCount {
+			newReplicasCount = maxCleanupCount - totalScaledDown
+		}
```

Comment:
I don't think the logic here is correct. Let's say we have only one old rc (available pods = 3, replicas = 5), and the maxCleanupCount = 1. After this call, we expect the rc to be scaled down to (available pods = 3, replicas = 4). 

However, based on the logic here, newReplicasCount becomes 1. Then we scale this rc down to replicas = 1, and thus available pods decreased. 

Also, can you make the logic here less complex? It's kind of hard to follow.